File: solutions/ruby/locomotive-engineer/2/locomotive_engineer.rb
class LocomotiveEngineer
  def self.generate_list_of_wagons(*args) 
    args
  end

  def self.fix_list_of_wagons(each_wagons_id, missing_wagons)
    misplaced_1, misplaced_2, locomotive, *other_wagons = each_wagons_id
    [locomotive, *missing_wagons, *other_wagons, misplaced_1, misplaced_2]
  end

  def self.add_missing_stops(route, **stops)
    {**route, stops: stops.values}
  end


  def self.extend_route_information(route, more_route_information)
    route.merge(more_route_information)
  end

  ## Alternate implementation - 
  # def self.extend_route_information(route, more_route_information)
  #   {**route, **more_route_information}
  # end
end

File: solutions/python/locomotive-engineer/2/locomotive_engineer.py
"""Functions which helps the locomotive engineer to keep track of the train."""


def get_list_of_wagons(*args) -> list:
    """Return a list of wagons.
    """
    return list(args)


def fix_list_of_wagons(each_wagons_id, missing_wagons):
    """Fix the list of wagons.
    """
    head, tail = each_wagons_id[:2], each_wagons_id[3:]
    return [1] + missing_wagons + tail + head

def add_missing_stops(routes, **kwargs):
    """Add missing stops to route dict.
    """
    routes['stops'] = list(kwargs.values())
    return routes


def extend_route_information(route, more_route_information):
    """Extend route information with more_route_information.
    """
    return {**route, **more_route_information}


def fix_wagon_depot(wagons_rows):
    """Fix the list of rows of wagons.
    """
    return [list(row) for row in zip(*wagons_rows)]
